ray_julia_jll/gen/make.jl

```diff
@@ -0,0 +1,168 @@
+    (?<=(?(<scheme>)/|:))
+    .*
+)?
+$
+"""x
+
+function create_tarball(dir, tarball)
+    return open(GzipCompressorStream, tarball, "w") do tar
+        Tar.create(dir, tar)
+    end
+end
+
+function remote_url(repo_root::AbstractString, name::AbstractString="origin")
+    return LibGit2.with(LibGit2.GitRepo(repo_root)) do repo
+        LibGit2.with(LibGit2.lookup_remote(repo, name)) do remote
+            LibGit2.url(remote)
+        end
+    end
+end
+
+function upload_to_github_release(
+    archive_path::AbstractString,
+    archive_url::AbstractString,
+    commit;
+    kwargs...
+)
+    return upload_to_github_release(archive_path, parse(URI, artifact_url), commit; kwargs...)
+end
+
+# TODO: Does this work properly with directories?
+function upload_to_github_release(archive_path::AbstractString, archive_uri::URI, commit; kwargs...)
+    # uri = parse(URI, artifact_url)
+    if archive_uri.host != "github.com"
+        throw(ArgumentError("Artifact URL is not for github.com: $(archive_uri)"))
+    end
+
+    m = match(GH_RELEASE_ASSET_PATH_REGEX, archive_uri.path)
+    if m === nothing
+        throw(ArgumentError(
+            "Artifact URL is not a GitHub release asset path: $(archive_uri)"
+        ))
+    end
+
+    # The `ghr` utility uses the local file name for the release asset. In order to have
+    # have the asset match the specified URL we'll temporarily rename the file.
+    org_archive_name = nothing
+    if basename(archive_path) != m[:file_name]
+        org_archive_name = basename(archive_path)
+        archive_path = mv(archive_path, joinpath(dirname(archive_path), m[:file_name]))
+    end
+
+    upload_to_github_release(m[:owner], m[:repo_name], commit, m[:tag], archive_path; kwargs...)
+
+    # Rename the archive back to the original name
+    if org_archive_name !== nothing
+        mv(archive_path, joinpath(dirname(archive_path), org_archive_name))
+    end
+
+    return nothing
+end
+
+
+function upload_to_github_release(owner, repo_name, commit, tag, path; token=ENV["GITHUB_TOKEN"])
+    # Based on: https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/d40ec617d131a1787851559ef1a9f04efce19f90/src/AutoBuild.jl#L487
+    # TODO: Passing in a directory path uploads multiple assets
+    # TODO: Would be nice to perform parallel uploads
+    cmd = ```
+        $(ghr()) \
+        -owner $owner \
+        -repository $repo_name \
+        -commitish $commit \
+        -token $token \
+        -n $tag \
+        -prerelease \
+        $tag $path
+    ```
+
+    run(cmd)
+end
+
+if abspath(PROGRAM_FILE) == @__FILE__
+
+    # Check for this now before we spend any time building the JLL
+    !haskey(ENV, "GITHUB_TOKEN") && error("\"GITHUB_TOKEN\" environment variable required.")
+
+    repo_path = abspath(joinpath(@__DIR__, "..", ".."))
+    pkg_url = remote_url(repo_path)
+
+    # Build JLL
+    @info "Building ray_julia_jll..."
+    Pkg.build("ray_julia_jll"; verbose=true)
```

Comment:
Once artifacts are the norm for this package we won't use a `deps/build.jl` script anymore. That code will still need to exist but `Pkg.build("ray_julia_jll")` shouldn't do anything.

Reply:
May want to break this up into a `gen/build.jl` and a `gen/upload.jl`

Reply:
Hmm...if we break it up then we'd have to pass `clean=false` to `mktempdir` or (probably better) create a local directory with a given name that we add to `.gitignore`.

Reply:
re `Pkg.build` I think I'll deprecate that as part of https://github.com/beacon-biosignals/Ray.jl/pull/134 since  the package won't be able to pull the JLLs until we've updated the Artifacts anyway...

Reply:
Works for me

Reply:
done